hs/src/Reach/Simulator/Core.hs

```diff
@@ -478,40 +460,32 @@ instance Interp DLExpr where
     DLE_MapRef _at dlmvar dlarg -> do
       (g, _) <- getState
       let linstate = e_linstate g
-      ev <- interp dlarg
-      case ev of
-        V_Address acc -> do
-          let m = saferMapRef "DLE_MapRef1" $ M.lookup dlmvar linstate
-          return $ saferMapRef "DLE_MapRef2" $ M.lookup acc m
-        _ -> impossible "unexpected error"
+      acc <- vAddress <$> interp dlarg
+      let m = saferMapRef "DLE_MapRef1" $ M.lookup dlmvar linstate
+      return $ saferMapRef "DLE_MapRef2" $ M.lookup acc m
     DLE_MapSet _at dlmvar dlarg maybe_dlarg -> do
       (e, _) <- getState
       let linst = e_linstate e
-      ev <- interp dlarg
-      case ev of
-        V_Address acc -> case maybe_dlarg of
-          Nothing -> do
-            let m = M.delete acc $ saferMapRef "DLE_MapSet1" $ M.lookup dlmvar linst
-            setGlobal $ e {e_linstate = M.insert dlmvar m linst}
-            return V_Null
-          Just dlarg' -> do
-            ev' <- interp dlarg'
-            let m = M.insert acc ev' $ saferMapRef "DLE_MapSet2" $ M.lookup dlmvar linst
-            setGlobal $ e {e_linstate = M.insert dlmvar m linst}
-            return V_Null
-        _ -> impossible "unexpected error"
+      acc <- vAddress <$> interp dlarg
+      case maybe_dlarg of
+        Nothing -> do
+          let m = M.delete acc $ saferMapRef "DLE_MapSet1" $ M.lookup dlmvar linst
+          setGlobal $ e {e_linstate = M.insert dlmvar m linst}
+          return V_Null
+        Just dlarg' -> do
+          ev' <- interp dlarg'
+          let m = M.insert acc ev' $ saferMapRef "DLE_MapSet2" $ M.lookup dlmvar linst
+          setGlobal $ e {e_linstate = M.insert dlmvar m linst}
```

Comment:
Really ugly; should combine into one thing where `maybe_dlarg` computes an update function:
```
f <- case maybe_dlarg of
  Nothing -> return M.delete
  Just a -> do
   v <- interp a
   return $ flip M.insert v
let m = f acc $ saferMapRef "DLE_MapSet" $ M.lookup dlmvar linst
setGlobal $ e {e_linstate = M.insert dlmvar m linst}
return V_Null
```